### PR TITLE
[WIP] Reload using magazines (part 6)

### PR DESCRIPTION
**Pull request by mugling**
This PR refactors `player::fire_gun()`

Depends upon ~~#14481~~, ~~#14482~~, #14487, ~~#14493~~ and ~~#14495~~ which are submitted separately to ease testing.

### CODE REVIEW

File: src/ranged.cpp [hunk 1]
```diff
@@ -443,38 +410,16 @@ void player::fire_gun( const tripoint &targ_arg, bool burst )
     // This is expensive, let's cache. todo: figure out if we need weapon.range(&p);
     const int weaponrange = used_weapon->gun_range( this );
 
-    // If the dispersion from the weapon is greater than the dispersion from your skill,
-    // you can't tell if you need to correct or the gun messed you up, so you can't learn.
-    const int weapon_dispersion = used_weapon->get_curammo()->ammo->dispersion +
```
Comment (mugling) [inline src/ranged.cpp]: Effectively the same as calling `item::gun_dispersion(true)`

**Comment by mugling** [comment]
This graph shows experience gained per shot via `player::practice` as currently calculated at the end of the loop and clearly shows the last two conditionals (`<= 0.6` and `<= 1.0`) are effectively unused.

sktrain

The magic constant `damage_factor` appears to be arbitrary and won't affect the shape of the curves. Does anyone know from where it was originally derived?

File: src/ranged.cpp [hunk 2]
```diff
@@ -405,15 +394,10 @@ void player::fire_gun( const tripoint &targ_arg, bool burst )
...
-        used_weapon->gun_dispersion(false);
-    const int player_dispersion = skill_dispersion( used_weapon, false ) +
-        ranged_skill_offset( used_weapon->gun_skill() );
-    // High perception allows you to pick out details better, low perception interferes.
-    ///\EFFECT_PER allows you to learn more often while using less accurate ranged weapons
```
Comment (kevingranade) [inline src/ranged.cpp]: This is a doxygen self-documenting comment, please preserve them.

Comment (sparr) [inline src/ranged.cpp]: I guess I need to write documentation for the doxygen documentation...

**Pull request merged**